components/gcp/container/component_sdk/python/kfp_component/google/dataflow/_launch_flex_template.py

```diff
@@ -0,0 +1,103 @@
+)
+
+
+def launch_flex_template(
+    project_id,
+    location,
+    launch_parameters,
+    validate_only=False,
+    staging_dir=None,
+    wait_interval=30,
+    job_id_output_path='/tmp/kfp/output/dataflow/job_id.txt',
+    job_object_output_path='/tmp/kfp/output/dataflow/job.json',
+):
+    """Launches a dataflow job from a flex template.
+
+    Args:
+        project_id (str): Required. The ID of the Cloud Platform project that the job belongs to.
+        location (str): The regional endpoint to which to direct the request.
+        launch_parameters (dict): Parameters to provide to the template
+            being launched. Schema defined in
+            https://cloud.google.com/dataflow/docs/reference/rest/v1b3/projects.locations.flexTemplates/launch#LaunchFlexTemplateParameter.
+            `jobName` will be replaced by generated name.
+        validate_only (boolean): If true, the request is validated but
+            not actually executed. Defaults to false.
+        staging_dir (str): Optional. The GCS directory for keeping staging files.
+            A random subdirectory will be created under the directory to keep job info
+            for resuming the job in case of failure.
+        wait_interval (int): The wait seconds between polling.
+        job_id_output_path (str): Optional. Output file to save job_id of execution
+        job_object_output_path (str): Optional. Output file to save job details of execution
+
+    Returns:
+        The completed job.
+    """
+    storage_client = storage.Client()
+    df_client = DataflowClient()
+    job_id = None
+
+    def cancel():
+        if job_id:
+            df_client.cancel_job(project_id, job_id, location)
+
+    with KfpExecutionContext(on_cancel=cancel) as ctx:
+        staging_location = get_staging_location(staging_dir, ctx.context_id())
+        job_id, _ = read_job_id_and_location(storage_client, staging_location)
+        # Continue waiting for the job if it's has been uploaded to staging location.
+        if job_id:
+            job = df_client.get_job(project_id, job_id, location)
+            job = wait_and_dump_job(
+                df_client,
+                project_id,
+                location,
+                job,
+                wait_interval,
+                job_id_output_path=job_id_output_path,
+                job_object_output_path=job_object_output_path,
+            )
+            logging.info(f'Skipping, existing job: {job}')
+            return job
+
+        if launch_parameters is None:
+            launch_parameters = {}
+
+        request_body = {
+            'launchParameter': launch_parameters,
+            'validateOnly': validate_only
+        }
+
+        request_body['launchParameter']['jobName'] = 'job-' + ctx.context_id()
+
+        response = df_client.launch_flex_template(
+            project_id, request_body, location
+        )
+
+        job = response.get('job', None)
+        if not job:
+            # Validate only mode
```

Comment:
For my education. What response you get when validation failed? In that case, should we return something orlog some error message?

Reply:
If the validation fails, then you get a HTTP 400 error from the underlying client and this will fail the step in the workflow.

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/local/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/ml/kfp_component/launcher/__main__.py", line 45, in <module>
    main()
  File "/ml/kfp_component/launcher/__main__.py", line 42, in main
    launch(args.file_or_module, args.args)
  File "/ml/kfp_component/launcher/launcher.py", line 53, in launch
    return fire.Fire(module, command=args, name=module.__name__)
  File "/usr/local/lib/python3.7/site-packages/fire/core.py", line 127, in Fire
    component_trace = _Fire(component, args, context, name)
  File "/usr/local/lib/python3.7/site-packages/fire/core.py", line 366, in _Fire
    component, remaining_args)
  File "/usr/local/lib/python3.7/site-packages/fire/core.py", line 542, in _CallCallable
    result = fn(*varargs, **kwargs)
  File "/ml/kfp_component/google/dataflow/_launch_flex_template.py", line 81, in launch_flex_template
    project_id, request_body, location
  File "/ml/kfp_component/google/dataflow/_client.py", line 37, in launch_flex_template
    projectId=project_id, location=location, body=request_body
  File "/usr/local/lib/python3.7/site-packages/googleapiclient/_helpers.py", line 130, in positional_wrapper
    return wrapped(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/googleapiclient/http.py", line 851, in execute
    raise HttpError(resp, content, uri=self.uri)
googleapiclient.errors.HttpError: <HttpError 400 when requesting 
https://dataflow.googleapis.com/v1b3/projects/<project_id_redacted>/locations/us-central1/flexTemplates:launch?alt=json
 returned "The template parameters are invalid.">
``` 

**Note:**  In case of validate only mode the step will fail even if the parameters are valid. This is because there is no real Dataflow job created and no information is saved to the output artifacts of the workflow step. 

```
This step is in Error state with this message: failed to save outputs: path /tmp/outputs/job_id/data does not exist (or /tmp/outputs/job_id/data is empty) in archive /argo/outputs/artifacts/launch-dataflow-flex-template-job_id.tgz
```

Personally, I think this flag is of less value the way it is currently implemented. The best use is during the development phase of the pipeline for someone to test the input arguments, but for all practical purposes the validate only option will be false. 
I added this for consistency with the existing classic Dataflow template component. Open to suggestions here.

Reply:
I ended up removing the `validate_only` parameter from the `component.yaml` because I didn't feel it added much value.